interfaces/apparmor/apparmor.go

```diff
@@ -0,0 +1,94 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor contains primitives for working with apparmor.
+//
+// References:
+//  - http://wiki.apparmor.net/index.php/Kernel_interfaces
+//  - http://apparmor.wiki.kernel.org/
+//  - http://manpages.ubuntu.com/manpages/xenial/en/man7/apparmor.7.html
+package apparmor
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// LoadProfile loads an apparmor profile from the given file.
+//
+// If no such profile was previously loaded then it is simply added to the kernel.
+// If there was a profile with the same name before, that profile is replaced.
+func LoadProfile(fname string) error {
+	return exec.Command("apparmor_parser", "--replace", fname).Run()
+}
+
+// Profile contains the name and mode of an apparmor profile loaded into the kernel.
+type Profile struct {
+	// Name of the profile. This is is either full path of the executable or an
+	// arbitrary string without spaces.
+	Name string
+	// Mode is either "enforcing" or "complaining".
+	Mode string
+}
+
+// Unload removes a profile from the running kernel.
+//
+// The operation is done with: apparmor_parser --remove $name
+func (profile *Profile) Unload() error {
+	return exec.Command("apparmor_parser", "--remove", profile.Name).Run()
+}
+
+// profilesPath contains information about the currently loaded apparmor profiles.
+const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
+
+var profilesPath = realProfilesPath
+
+// LoadedProfiles interrogates the kernel and returns a list of loaded apparmor profiles.
+//
+// Snappy manages apparmor profiles named *.snap. Other profiles might exist on
```

Comment:
On Fri, 2016-03-11 at 00:47 -0800, Zygmunt Krynicki wrote:
> > 
> > +
> > +// Unload removes a profile from the running kernel.
> > +//
> > +// The operation is done with: apparmor_parser --remove $name
> > +func (profile *Profile) Unload() error {
> > +	return exec.Command("apparmor_parser", "--remove",
> > profile.Name).Run()
> > +}
> > +
> > +// profilesPath contains information about the currently loaded
> > apparmor profiles.
> > +const realProfilesPath = "/sys/kernel/security/apparmor/profiles"
> > +
> > +var profilesPath = realProfilesPath
> > +
> > +// LoadedProfiles interrogates the kernel and returns a list of
> > loaded apparmor profiles.
> > +//
> > +// Snappy manages apparmor profiles named *.snap. Other profiles
> > might exist on
> But as mentioned the chosen implementation has a flaw with 'oh.snap'.
> Snappy will think it is managing oh.snap when it isn't.
> 
> ZK: We need a compromise. There are no "namespaces" for apparmor
> profiles. We're broadly allocating ``*.snap`` in order to make one.
> Unless there are existing "oh.snap" profiles in the wild I think this
> is exactly what we should do, in the absence of better kernel
> resource management, to send a strong signal that snappy is using
> apparmor in a certain way and that you should not name your profiles
> ``*.snap``
> 
I discussed this with the security team and none of us like the
approach of appending ".snap". I think the compromise is looking at the
filesystem in the manner I described previously instead of using a
string to attempt namespacing. 

AppArmor will have the concept of policy namespaces in support of LXD
for 16.04. The feature hasn't landed yet and is coming in hot. I
discussed this feature at length with the AppArmor team and I don't
advise using it for 16.04 because there are many security policy
implications to using them (ie, it is far more than just creating a
namespace and loading into it) and we are going to have to properly
design their use in snappy.

-- 
Jamie Strandboge             | http://www.canonical.com